src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,152 @@
+            head = newNode;
+        } else {
+            predecessor.next = newNode;
+        }
+        size++;
+    }
+
+    private void linkToTail(T elementToInsert) {
+        Node<T> predecessor = this.tail;
+        Node<T> newNode = new Node<>(predecessor, elementToInsert, null);
+        this.tail = newNode;
+        if (predecessor == null) {
+            head = newNode;
+        } else {
+            predecessor.next = newNode;
+        }
+        size++;
+    }
+
+    private Node<T> getNodeByIndex(int index) {
+        Node<T> targetNode = head;
+        for (int i = 0; i < index; i++) {
+            targetNode = targetNode.next;
+        }
+        return targetNode;
+    }
+
+    private T removeNode(Node<T> nodeToRemove) {
+        Node<T> beforeTarget = nodeToRemove.prev;
+        Node<T> afterTarget = nodeToRemove.next;
+        if (afterTarget == null) {
+            tail = beforeTarget;
+            nodeToRemove.prev = null;
+        } else if (beforeTarget == null) {
+            head = afterTarget;
+            nodeToRemove.next = null;
+        } else {
+            beforeTarget.next = afterTarget;
+            afterTarget.prev = beforeTarget;
+        }
+        size--;
+        return nodeToRemove.value;
+    }
+
     @Override
     public boolean add(T value) {
-        return false;
+        linkToTail(value);
+        return true;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index == size) {
+            linkToTail(value);
+        } else if (index < size && index >= 0) {
+            linkToNode(value, getNodeByIndex(index));
+        } else {
+            throw new IndexOutOfBoundsException(index + " index is out of bounds");
+        }
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        for (T item: list) {
+            add(item);
+        }
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (isInvalidIndex(index)) {
+            throw new IndexOutOfBoundsException(index + " index is out of bounds");
+        }
+        return getNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        if (isInvalidIndex(index)) {
+            throw new IndexOutOfBoundsException(index + " index is out of bounds");
+        }
+        Node<T> oldElement = getNodeByIndex(index);
+        T oldNodeValue = oldElement.value;
+        oldElement.value = value;
+        return oldNodeValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (isInvalidIndex(index)) {
+            throw new IndexOutOfBoundsException(index + " index is out of bounds");
+        }
+        return removeNode(getNodeByIndex(index));
     }
 
     @Override
     public boolean remove(T object) {
+        for (int i = 0; i < size; i++) {
+            T targetValue = getNodeByIndex(i).value;
```

Comment:
you use nested loop (one here, second in getNodeByIndex) can we do it with single loop? 
that decrease algorithmic complexity

Reply:
used `while` loop instead, checking until `node` is `null`